### PR TITLE
HBASE-22817 Use hbase-shaded dependencies in hbase-spark

### PR DESCRIPTION
**Pull request by meszibalu**
* hbase-spark-protocol: a new project for the protobuf files.
* hbase-spark-protocol-shaded: the shaded version of hbase-spark-protocol,
  which is compatible with hbase thirdparty protobuf.

hbase-spark codebase now depends on only hbase-shaded-* stuff and
hbase-spark-protocol-shaded.

### CODE REVIEW

**Comment by meszibalu** [comment]
retest build

**Comment by busbey** [comment]
looking at this locally maybe HBASE-23059 is sufficient to correct? can you rebase on top of the latest master?

**Comment by meszibalu** [comment]
I did the rebase but it still fails. This package is definitely in hbase-spark-protocol-shaded package.

**Comment by busbey** [comment]
okay let me dig in and see what's going on.

**Comment by busbey** [comment]
Okay I figured this out.

the easiest way to reproduce the problem is to just go to the top level and do `mvn --batch-mode clean test-compile -DskipTests=true`

the problem is this bit in hbase-spark

```

    <dependency>
      <groupId>org.apache.hbase.connectors.spark</groupId>
      <artifactId>hbase-spark-protocol-shaded</artifactId>
      <exclusions>
        <exclusion>
          <groupId>org.apache.hbase.connectors.spark</groupId>
          <artifactId>hbase-spark-protocol</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
```

So long as we run though a `package` phase this does the correct thing. We get a dependency on the classes in `hbase-spark-protocol-shaded` that have our generated classes from `hbase-spark-protocol` rewritten to use the shaded protobuf and avoid an extra copy of those classes without the rewrite.

The problem presents when we don't do a package phase. At that point maven relies on the classes output by modules included in a reactor build (rather than e.g. jars from a prior installed build) and if we look at the `hbase-spark-protocol-shaded` module we'll see there are none:

```
Busbey-MBA:hbase-connectors busbey$ find spark/hbase-spark-protocol-shaded/target/
spark/hbase-spark-protocol-shaded/target/
spark/hbase-spark-protocol-shaded/target//test-classes
spark/hbase-spark-protocol-shaded/target//test-classes/META-INF
spark/hbase-spark-protocol-shaded/target//test-classes/META-INF/LICENSE
spark/hbase-spark-protocol-shaded/target//test-classes/META-INF/DEPENDENCIES
spark/hbase-spark-protocol-shaded/target//test-classes/META-INF/NOTICE
spark/hbase-spark-protocol-shaded/target//maven-shared-archive-resources
spark/hbase-spark-protocol-shaded/target//maven-shared-archive-resources/META-INF
spark/hbase-spark-protocol-shaded/target//maven-shared-archive-resources/META-INF/LICENSE
spark/hbase-spark-protocol-shaded/target//maven-shared-archive-resources/META-INF/DEPENDENCIES
spark/hbase-spark-protocol-shaded/target//maven-shared-archive-resources/META-INF/NOTICE
spark/hbase-spark-protocol-shaded/target//classes
spark/hbase-spark-protocol-shaded/target//classes/META-INF
spark/hbase-spark-protocol-shaded/target//classes/META-INF/LICENSE
spark/hbase-spark-protocol-shaded/target//classes/META-INF/DEPENDENCIES
spark/hbase-spark-protocol-shaded/target//classes/META-INF/NOTICE
spark/hbase-spark-protocol-shaded/target//.plxarc
Busbey-MBA:hbase-connectors busbey$ 
```

**Comment by busbey** [comment]
we can either fix this by making our build/test time rely on the unshaded version of `hbase-spark-protocol` (which only works if we don't need to reference any of the protobuf stuff) or we can update our personality to ensure a maven package phase runs for these tests.

**Comment by meszibalu** [comment]
According to `maven-branch-dirlist-root.txt`, yetus fails at the beginning:

`[ERROR] Failed to execute goal on project hbase-kafka-proxy: Could not resolve dependencies for project org.apache.hbase.connectors.kafka:hbase-kafka-proxy:jar:1.0.1-SNAPSHOT: Could not find artifact org.apache.hbase.connectors.kafka:hbase-kafka-model:jar:1.0.1-SNAPSHOT in Nexus (http://repository.apache.org/snapshots) -> [Help 1]
[ERROR] Failed to execute goal on project hbase-spark-it: Could not resolve dependencies for project org.apache.hbase.connectors.spark:hbase-spark-it:jar:1.0.1-SNAPSHOT: Could not find artifact org.apache.hbase.connectors.spark:hbase-spark:jar:1.0.1-SNAPSHOT in Nexus (http://repository.apache.org/snapshots) -> [Help 1]`

**Comment by meszibalu** [comment]
Which leads to:
`[Fri Sep 27 13:30:06 CEST 2019 DEBUG]: Maven: finish re-ordering modules
[Fri Sep 27 13:30:06 CEST 2019 DEBUG]: Finished list: kafka kafka/hbase-kafka-model kafka/hbase-kafka-proxy spark spark/hbase-spark-protocol hbase-connectors-assembly . spark/hbase-spark spark/hbase-spark-it spark/hbase-spark-protocol-shaded`

So hbase-spark-protocol* stuff is at the and of CHANGED_MODULES.

**Comment by busbey** [comment]
I've pushed a go at getting things working under Yetus. AFAICT this works in my local checkout. I have not yet tried it with docker, so I'm pushing here to see how it goes.

File: pom.xml [hunk 1]
```diff
@@ -491,7 +514,7 @@
         <artifactId>flatten-maven-plugin</artifactId>
         <version>1.1.0</version>
         <configuration>
-          <updatePomFile>true</updatePomFile>
+          <updatePomFile>false</updatePomFile>
```
Comment (busbey) [inline pom.xml]: looking at this more locally, I think setting this to false also made it so the plugin didn't put an updated pom into the repo as a result of running `install`. It also looks like after the changes to wrap the maven plugin for yetus we don't generate maven commands that cause the flatten plugin to fail with this set to true.

**Comment by meszibalu** [comment]
Thanks @busbey, it compiles for me locally, let me push it.

**Pull request merged**